File: android/src/main/java/com/margelo/nitro/googleplaces/HybridPlaces.kt
package com.margelo.nitro.googleplaces

import android.util.Log
import com.margelo.nitro.NitroModules
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.CircularBounds
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.RectangularBounds

import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.PlacesClient
import com.margelo.nitro.core.NullType
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.tasks.await

class HybridPlaces() : HybridPlacesSpec() {

    private val placesClient: PlacesClient

    // Accessed only inside synchronized(sessionTokens) in sessionTokenFor, so a
    // plain map is enough (getOrPut + remove is a compound op that a
    // synchronizedMap would not make atomic anyway). Insertion-ordered so the
    // eldest abandoned session is evicted once the cap is reached.
    private val sessionTokens =
        object : LinkedHashMap<String, AutocompleteSessionToken>() {
            override fun removeEldestEntry(
                eldest: MutableMap.MutableEntry<String, AutocompleteSessionToken>
            ): Boolean = size > MAX_SESSION_TOKENS
        }

    private fun sessionTokenFor(sessionId: String?, consume: Boolean): AutocompleteSessionToken {
        if (sessionId == null) {
            return AutocompleteSessionToken.newInstance()
        }
        synchronized(sessionTokens) {
            val token = sessionTokens.getOrPut(sessionId) { AutocompleteSessionToken.newInstance() }
            if (consume) {
                sessionTokens.remove(sessionId)
            }
            return token
        }
    }

    init {
        val context =
            NitroModules.applicationContext
                ?: throw IllegalStateException("React Context is not initialized")
        // Recupera l'API Key dal Manifest (meta-data) o stringhe
        val apiKey = context.packageManager
            .getApplicationInfo(context.packageName, android.content.pm.PackageManager.GET_META_DATA)
            .metaData.getString("com.google.android.geo.API_KEY")

        if (apiKey != null) {
            if (!Places.isInitialized()) {
                Places.initializeWithNewPlacesApiEnabled(context, apiKey)
            }
            println("[HybridPlaces] Places SDK initialized successfully.")
        } else {
            println("[HybridPlaces] ERROR: Missing API Key in AndroidManifest.xml")
        }
        
        placesClient = Places.createClient(context)
    }

    override fun autocomplete(query: String, options: AutocompleteOptions?): Promise<Array<PlaceAutocompleteResult>> {
        return Promise.async {
            val token = sessionTokenFor(options?.sessionToken, consume = false)
            val effectiveTypes = options?.types?.toList() ?: listOf("route", "street_address", "premise", "subpremise", "geocode")
            val requestBuilder = FindAutocompletePredictionsRequest.builder()
                .setSessionToken(token)
                .setQuery(query)
                .setTypesFilter(effectiveTypes)

            options?.countries?.let { requestBuilder.setCountries(it.toList()) }

            options?.locationBias?.let { bias ->
                bias.match(
                    first = { circular ->
                        requestBuilder.setLocationBias(
                            CircularBounds.newInstance(
                                LatLng(circular.latitude, circular.longitude),
                                circular.radius
                            )
                        )
                    },
                    second = { rectangular ->
                        requestBuilder.setLocationBias(
                            RectangularBounds.newInstance(
                                LatLng(rectangular.southWestLatitude, rectangular.southWestLongitude),
                                LatLng(rectangular.northEastLatitude, rectangular.northEastLongitude)
                            )
                        )
                    }
                )
            }

            options?.locationRestriction?.let { restriction ->
                restriction.match(
                    first = { circular ->
                        requestBuilder.setLocationRestriction(
                            CircularBounds.newInstance(
                                LatLng(circular.latitude, circular.longitude),
                                circular.radius
                            )
                        )
                    },
                    second = { rectangular ->
                        requestBuilder.setLocationRestriction(
                            RectangularBounds.newInstance(
                                LatLng(rectangular.southWestLatitude, rectangular.southWestLongitude),
                                LatLng(rectangular.northEastLatitude, rectangular.northEastLongitude)
                            )
                        )
                    }
                )
            }

            val response = placesClient.findAutocompletePredictions(requestBuilder.build()).await()
            
            response.getAutocompletePredictions().map { suggestion ->
                PlaceAutocompleteResult(
                    suggestion.placeId,
                    suggestion.getFullText(null).toString()
                )
            }.toTypedArray()
        }
    }

    override fun getPlace(placeId: String, options: GetPlaceOptions?): Promise<Variant_NullType_PlaceDetails> {
        return Promise.async {
            try {
                val placeFields = listOf(
                    Place.Field.ID,
                    Place.Field.DISPLAY_NAME,
                    Place.Field.FORMATTED_ADDRESS,
                    Place.Field.ADDRESS_COMPONENTS,
                    Place.Field.LOCATION
                )

                val request = FetchPlaceRequest.builder(placeId, placeFields)
                    .setSessionToken(sessionTokenFor(options?.sessionToken, consume = true))
                    .build()

                val response = placesClient.fetchPlace(request).await()
                val place = response.place

                val components = place.addressComponents?.asList()?.map { component ->
                    AddressComponent(
                        component.name,
                        component.shortName ?: "",
                        component.types.toTypedArray()
                    )
                }?.toTypedArray() ?: emptyArray()

                val details = PlaceDetails(
                    place.displayName ?: "",
                    place.formattedAddress ?: "",
                    place.id ?: "",
                    place.location?.latitude ?: 0.0,
                    place.location?.longitude ?: 0.0,
                    components
                )

                Variant_NullType_PlaceDetails.create(details)
            } catch (e: Exception) {
                Log.e("HybridPlaces", "getPlace error: ${e.message}")
                Variant_NullType_PlaceDetails.create(NullType.NULL)
            }
        }
    }

    override fun autocompleteWithDetails(query: String, options: AutocompleteOptions?): Promise<Array<PlaceDetails>> {
        return Promise.async {
            val predictions = autocomplete(query, options).await()
            val detailedResults = predictions.mapIndexedNotNull { index, prediction ->
                try {
                    // Only the first detail fetch closes the shared session; the
                    // rest use their own single-use tokens so we don't reuse a
                    // consumed one.
                    val detailToken = if (index == 0) options?.sessionToken else null
                    val variant = getPlace(prediction.placeId, GetPlaceOptions(detailToken)).await()
                    variant.asSecondOrNull()
                } catch (e: Exception) {
                    null
                }
            }
            detailedResults.toTypedArray()
        }
    }

    companion object {
        private const val MAX_SESSION_TOKENS = 100
    }
}

File: ios/HybridPlaces.swift
import Foundation
import CoreLocation
import NitroModules
import GooglePlaces


class HybridPlaces : HybridPlacesSpec {
      /// Native session tokens keyed by the caller-provided session id, so a
      /// single search reuses one token across keystrokes instead of minting a
      /// fresh (separately billed) token on every call. Accessed only on the
      /// main queue, matching where the Places SDK requests are issued.
      private var sessionTokens: [String: GMSAutocompleteSessionToken] = [:]
      /// Guards against unbounded growth from abandoned searches (a session that
      /// never reaches `getPlace`). Tokens also expire server-side after a few
      /// minutes, so evicting the oldest here is safe.
      private var sessionTokenOrder: [String] = []
      private let maxSessionTokens = 100

      override init() {
          super.init()
        DispatchQueue.main.async {
                if let apiKey = Bundle.main.object(forInfoDictionaryKey: "GMSPlacesAPIKey") as? String {
                    GMSPlacesClient.provideAPIKey(apiKey)
                    print("[HybridPlaces] GMSPlacesClient API key provided successfully.")
                } else {
                    print("[HybridPlaces] ERROR: Missing GMSPlacesAPIKey in Info.plist")
                }
            }
      }

    /// Returns the session token to use for a request. When `sessionId` is nil a
    /// fresh single-use token is returned (legacy behaviour). Otherwise the token
    /// for that id is reused, creating one on first use. When `consume` is true
    /// (a `getPlace` call) the token is removed afterwards, closing the session.
    private func sessionToken(for sessionId: String?, consume: Bool) -> GMSAutocompleteSessionToken {
      guard let sessionId = sessionId else {
        return GMSAutocompleteSessionToken()
      }

      let token: GMSAutocompleteSessionToken
      if let existing = sessionTokens[sessionId] {
        token = existing
      } else {
        token = GMSAutocompleteSessionToken()
        sessionTokens[sessionId] = token
        sessionTokenOrder.append(sessionId)
        if sessionTokenOrder.count > maxSessionTokens {
          let oldest = sessionTokenOrder.removeFirst()
          sessionTokens.removeValue(forKey: oldest)
        }
      }

      if consume {
        sessionTokens.removeValue(forKey: sessionId)
        sessionTokenOrder.removeAll { $0 == sessionId }
      }

      return token
    }

    func autocomplete(query: String, options: AutocompleteOptions?) throws -> Promise<[PlaceAutocompleteResult]> {
      return Promise.async {
        try await withCheckedThrowingContinuation { continuation in
          DispatchQueue.main.async {
            let token = self.sessionToken(for: options?.sessionToken, consume: false)
            let filter = GMSAutocompleteFilter()
            filter.types = options?.types ?? ["route", "street_address", "premise", "subpremise", "geocode"]
            if let countries = options?.countries {
              filter.countries = countries
            }
            func toLocationOption(
              _ bounds: Variant_CircularLocationBounds_RectangularLocationBounds
            ) -> any GMSPlaceLocationBias & GMSPlaceLocationRestriction {
              switch bounds {
              case .first(let circular):
                let center = CLLocationCoordinate2D(
                  latitude: circular.latitude,
                  longitude: circular.longitude
                )
                return GMSPlaceCircularLocationOption(center, circular.radius)
              case .second(let rectangular):
                let northEast = CLLocationCoordinate2D(
                  latitude: rectangular.northEastLatitude,
                  longitude: rectangular.northEastLongitude
                )
                let southWest = CLLocationCoordinate2D(
                  latitude: rectangular.southWestLatitude,
                  longitude: rectangular.southWestLongitude
                )
                return GMSPlaceRectangularLocationOption(northEast, southWest)
              }
            }

            if let locationBias = options?.locationBias {
              filter.locationBias = toLocationOption(locationBias)
            }
            if let locationRestriction = options?.locationRestriction {
              filter.locationRestriction = toLocationOption(locationRestriction)
            }

            let request = GMSAutocompleteRequest(query: query)
            request.filter = filter
            request.sessionToken = token

            GMSPlacesClient.shared().fetchAutocompleteSuggestions(
              from: request,
              callback: { (results, error) in

              if let error = error {
                continuation.resume(
                  throwing: RuntimeError.error(
                    withMessage: error.localizedDescription
                  )
                )
                return
              }
                
              let mappedResults: [PlaceAutocompleteResult] =
                results?.reduce(into: []) { (acc, prediction) in
                  if let suggestion = prediction.placeSuggestion {
                    let item = PlaceAutocompleteResult(
                      placeId: suggestion.placeID,
                      label: suggestion.attributedFullText.string
                    )
                    acc.append(item)
                  }
                } ?? []

              continuation.resume(returning: mappedResults)
            })
          }
        }
      }
    }

    func getPlace(placeId: String, options: GetPlaceOptions?) throws -> Promise<Variant_NullType_PlaceDetails> {
      return Promise.async {
        try await withCheckedThrowingContinuation { continuation in
          DispatchQueue.main.async {
            let myProperties = [
                GMSPlaceProperty.placeID,
                GMSPlaceProperty.name,
                GMSPlaceProperty.formattedAddress,
                GMSPlaceProperty.addressComponents,
                GMSPlaceProperty.coordinate
            ].map {$0.rawValue}
            let token = self.sessionToken(for: options?.sessionToken, consume: true)
            let fetchPlaceRequest = GMSFetchPlaceRequest(placeID: placeId, placeProperties: myProperties, sessionToken: token)
                    
            GMSPlacesClient.shared().fetchPlace(with: fetchPlaceRequest) { place, error in
              if let error = error as NSError? {
                if error.domain == kGMSPlacesErrorDomain && error.code == 400 {
                    continuation.resume(returning: .first(NullType.null))
                    return
                }
                
                continuation.resume(
                  throwing: error
                )
                return
              }

              if let place = place {
                  let components = place.addressComponents?.map { component in
                      AddressComponent(
                          name: component.name,
                          short_name: component.shortName ?? "",
                          types: component.types
                      )
                  } ?? []
                  
                  let details = PlaceDetails(
                      name: place.name ?? "",
                      formatted_address: place.formattedAddress ?? "",
                      place_id: place.placeID ?? "",
                      latitude: place.coordinate.latitude,
                      longitude: place.coordinate.longitude,
                      address_components: components
                  )
                  
                  continuation.resume(returning: .second(details))
              } else {
                  continuation.resume(returning: .first(NullType.null))
              }
              
            }
          }
        }
      }
    }

    func autocompleteWithDetails(query: String, options: AutocompleteOptions?) throws -> Promise<[PlaceDetails]> {
        return Promise.async {
            let predictions = try await self.autocomplete(query: query, options: options).await()
            
            var detailedResults: [PlaceDetails] = []
            
            for (index, prediction) in predictions.enumerated() {
                // Only the first detail fetch closes the shared session; the rest
                // use their own single-use tokens so we don't reuse a consumed one.
                let detailToken = index == 0 ? options?.sessionToken : nil
                let placeResult = try await self.getPlace(placeId: prediction.placeId, options: GetPlaceOptions(sessionToken: detailToken)).await()
                
                if case .second(let details) = placeResult {
                  detailedResults.append(details)
                }
            }
            
            return detailedResults
        }
    }
}